compiler/rustc_hir_typeck/src/demand.rs

```diff
@@ -133,21 +133,26 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         expected: Ty<'tcx>,
         actual: Ty<'tcx>,
     ) -> Option<DiagnosticBuilder<'tcx, ErrorGuaranteed>> {
-        self.demand_eqtype_with_origin(&self.misc(sp), expected, actual)
+        self.demand_eqtype_with_origin(&self.misc(sp), expected, actual, |_| {})
     }
 
     pub fn demand_eqtype_with_origin(
         &self,
         cause: &ObligationCause<'tcx>,
         expected: Ty<'tcx>,
         actual: Ty<'tcx>,
+        mutate_err: impl FnOnce(&mut Diagnostic),
     ) -> Option<DiagnosticBuilder<'tcx, ErrorGuaranteed>> {
         match self.at(cause, self.param_env).eq(expected, actual) {
             Ok(InferOk { obligations, value: () }) => {
                 self.register_predicates(obligations);
                 None
             }
-            Err(e) => Some(self.err_ctxt().report_mismatched_types(cause, expected, actual, e)),
+            Err(e) => {
+                let mut diag = self.err_ctxt().report_mismatched_types(cause, expected, actual, e);
+                mutate_err(&mut diag);
+                Some(diag)
+            }
```

Comment:
Why add a closure, and not have the caller modify the returned value?

Reply:
I had originally thought to pass in the TypeError that resulted from the demand call, but then I didn't need it anymore, so this is just an artifact of that old design.